### PR TITLE
Issue 12 adding token information by --token-usage & -tu

### LINKED ISSUE #12: Get token information from Groq.
Based on the Groq instances, it would be a good idea to provide users with an option `--token_usage` or `-t` to monitor how many tokens are being returned or used for optimization purposes. This is important because some blocks of code might exceed the token limit applied to each model.

### PR DESCRIPTION
**Pull request by fadingNA**
# Description

Adding handleTokenFlag to check argument passing that have --token-usage or -tu if yes we stderr for the token usage.

- Build Succeed
Screenshot 2024-09-15 at 3 25 45 PM

- Tested
Screenshot 2024-09-15 at 3 25 19 PM

- Lintinting
Screenshot 2024-09-15 at 3 24 56 PM

# Checklist

- [x] Build does not fail.

- [x] Tested locally.

- [x] Linting errors if any are resolved.

# Testing Instructions

### 1. Create a 
- new dependencies `npm install`

### 2. Go to
- go to Optimizeit `cd OptimizeIt`
- npm run dev -- ./examples/dynamicMemory.cpp --token-usage

### CODE REVIEW

File: src/args/flag_handlers/tokenInfo.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+/**
+ * Handles the token info from Groq.
```
Comment (Mounayer) [inline src/args/flag_handlers/tokenInfo.ts]: suggestion: Maybe this can be more concise letting the reader know what the purpose of this is. i.e.: This flag dictates that information regarding the `chat prompt token count` and `chat completion token count` should be displayed.

File: src/args/flag_handlers/tokenInfo.ts [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+ *
+ * @param {string[]} args - The arguments passed to the CLI.
+ * @returns {boolean} True if the token usage flag is present, otherwise false.
+ */
+function handleTokenFlag(args: string[]): boolean {
+  try {
```
Comment (Mounayer) [inline src/args/flag_handlers/tokenInfo.ts]: suggestion: the try/catch block here is unnecessary, because no error is going to be thrown.

This should be something like:

```typescript
function handleTokenFlag(args: string[]): boolean {
return args.includes('--token-usage') || args.includes('-tu');
}
```

File: src/groq/index.ts [hunk 3]
```diff
@@ -111,4 +119,26 @@ export default class GroqChat {
...
+
+  /**
+   * Logs the token usage information to the console. by args --token-usage or -tu
+   *
+   * @param tokenUsageInfo - The token usage information object.
+   */
+  public saveTokenUsageInfo(tokenUsageInfo: Groq.CompletionUsage): void {
+    console.log('Token Usage Information:\n');
+
+    // Available properties:
```
Comment (Mounayer) [inline src/groq/index.ts]: suggestion: These comments should be removed as they are not necessary

File: src/groq/index.ts [hunk 4]
```diff
@@ -94,6 +95,13 @@ export default class GroqChat {
         temperature,
       );
 
+      if (tokenUsageInformation) {
+        if (!chatCompletion.usage) {
+          throw new Error('Token usage information is not available.');
+        }
+        this.saveTokenUsageInfo(chatCompletion?.usage);
```
Comment (Mounayer) [inline src/groq/index.ts]: suggestion: At the moment, we display the `Completion, Usage, and Total` tokens for each individual file processed. This can be cumbersome to read when `optimizeit` is processing several files at once. I'll leave it up to you, but, do you think a better approach here is to `accumulate` them and display the `completion, usage, and total` for all of the files processed instead of 1 by 1?

File: src/groq/index.ts [hunk 5]
```diff
@@ -111,4 +119,26 @@ export default class GroqChat {
...
+    //console.error(`queue_time: ${tokenUsageInfo.queue_time}`);
+    //console.error(`usage_prompt_tokens: ${tokenUsageInfo.prompt_tokens}`);
+    //console.error(`prompt_time: ${tokenUsageInfo.prompt_time}`);
+    //console.error(`completion_tokens: ${tokenUsageInfo.completion_tokens}`);
+    //console.error(`completion_time: ${tokenUsageInfo.completion_time}`);
+    //console.error(`total_time: ${tokenUsageInfo.total_time}`);
+    //console.error(`total_tokens: ${tokenUsageInfo.total_tokens}`);
+
+    console.error(`Completion Tokens: ${tokenUsageInfo.completion_tokens}`);
+    console.error(`Usage Prompt Tokens: ${tokenUsageInfo.prompt_tokens}`);
```
Comment (Mounayer) [inline src/groq/index.ts]: issue: Right now, these display above the output of every file processed, i.e.:

image

When `optimizeit` processes several files at once the output can be very big. This would then force the user to scroll all the way up to see the tokens which is not user friendly. I think that the token info needs to be displayed at the `bottom most` of the program output below all processed file outputs, i.e.:

image

**Review by Mounayer — CHANGES_REQUESTED** [review]
praise: LGTM! 🔥!! Looks great, well done, tested and works as expected, I just had a few suggestions and one minor issue for user experience.

File: src/groq/index.ts (hunk 3 above)
Comment (Mounayer) [reply src/groq/index.ts]: suggestion: comments still remain here, they should be removed, this function should also be `private` not `public` since it's only being used in this class

File: src/groq/index.ts [hunk 6]
```diff
@@ -111,4 +125,38 @@ export default class GroqChat {
...
+    this.totalTokens += tokenUsageInfo.total_tokens;
+  }
+
+  /**
+   * Logs the total token usage information to the console.
+   */
+  public logTotalTokenUsage(fileNames: string[]): void {
+    if (fileNames && fileNames.length > 0) {
+      for (const fileName of fileNames) console.log(`File: ${fileName}`);
+      console.log('\nHas the following token usage information:\n');
```
Comment (Mounayer) [inline src/groq/index.ts]: issue: this should be `stderr` i.e.: `console.error` not `console.log`

**Review by Mounayer — CHANGES_REQUESTED** [review]
praise: Great job addressing my comments earlier on, this is almost great! I just discovered one bug.

When I run `optimizeit` without specifying any `flags`, I get the following crash:

image


And just a few more suggestions!

File: src/main.ts [hunk 7]
```diff
@@ -30,16 +31,14 @@ async function main() {
 
   for (let i = 0; i < fileNames.length; ++i) {
     const filePath = path.resolve(fileNames[i]);
-
     const data = fileParser(filePath);
 
-    console.log(`\nOptimizing ${fileNames[i]}...\n`);
```
Comment (Mounayer) [inline src/main.ts:36]: question: why did you remove this?

File: src/groq/index.ts [hunk 8]
```diff
@@ -111,4 +124,38 @@ export default class GroqChat {
       }
     }
   }
+
+  /**
+   * Logs the token usage information to the console. by args --token-usage or -tu
+   *
+   * @param tokenUsageInfo - The token usage information object.
+   */
+  public accumulateToken(tokenUsageInfo: Groq.CompletionUsage): void {
```
Comment (Mounayer) [inline src/groq/index.ts]: suggestion: this function should be `private` because it's only being used in this class -- it should also be moved up because standard says:
- Private functions first
- Public functions second

File: src/groq/index.ts [hunk 9]
```diff
@@ -111,4 +124,38 @@ export default class GroqChat {
...
+    if (fileNames && fileNames.length > 0) {
+      for (const fileName of fileNames) console.error(`File: ${fileName}`);
+      console.error('\nHas the following token usage information:\n');
+    }
+
+    // Available properties:
+    //console.error(`queue_time: ${tokenUsageInfo.queue_time}`);
+    //console.error(`usage_prompt_tokens: ${tokenUsageInfo.prompt_tokens}`);
+    //console.error(`prompt_time: ${tokenUsageInfo.prompt_time}`);
+    //console.error(`completion_tokens: ${tokenUsageInfo.completion_tokens}`);
```
Comment (Mounayer) [inline src/groq/index.ts]: suggestion: all of the comments should be removed

**Review by Mounayer — CHANGES_REQUESTED** [review]
praise: LGTM!!! 🔥 🚀 

It's looking great now, just a few minor adjustments, mostly removing comments, or bringing back 1 comment you've deleted (probably accidentally I assume). And updating the README.md to include the newest added flag!

File: src/main.ts (hunk 7 above)
Comment (fadingNA) [reply src/main.ts:36]: Accidently removed sorry.

File: README.md [hunk 10]
```diff
@@ -141,6 +141,16 @@ optimizeit --help
 optimizeit -h
 ```
 
+### Token Usage
+
+This flag allows you print the completion token, prompt token, and total token that OptimizeIt is using from your prompt, usage:
+
+```bash
+optimizeit file1 --token-usage
+optimizeit file1 file2 --tu
```
Comment (Mounayer) [inline README.md]: nit: there's a small typo here, it's supposed to be `-tu` not `--tu`

**Review by Mounayer — CHANGES_REQUESTED** [review]
praise: LGTM!!! Thanks for updating `README.md` and the `-h` flag handler! It's looking great, just one tiny typo in `README.md` and it's good to go!!!

**Review by Mounayer — APPROVED** [review]
praise: LGTM! 🔥 🚀 !!! Well done, great job implementing this feature, it's a great addition to `optimizeit`, thank you for committing to this -- pun intended!

File: README.md [hunk 11]
```diff
@@ -141,6 +141,16 @@ optimizeit --help
 optimizeit -h
 ```
 
+### Token Usage
+
+This flag allows you print the completion token, prompt token, and total token that OptimizeIt is using from your prompt, usage:
+
+```bash
+optimizeit file1 --token-usage
+optimizeit file1 file2 -tu
```
Comment (humphd) [inline README.md:150]: NOTE: you typically want to use a single arg for `-` vs. `--` when doing flags.  I'd suggest `-u`.  A few CLIs break this rule, but most Unix tools follow it.

Comment (Mounayer) [reply README.md:150]: Oh interesting! I didn't know that, thanks for pointing it out
